### PR TITLE
Making the RD Servers do something than take up space.

### PR DESCRIPTION
**Pull request by WarlockD**
## About The Pull Request
My OCD finally couldn't take it anymore.  The RD Servers has code, its old but its there, that was supposed to handle techweb point generation.  If they overheat, they would to slow down point generation till someone turned the freezer back on.  Alas all of that code was skipped over in favor of a hard constant in the research subsystem.  Before this PR, as long as one server was somewhere in the world, point generation would continue.

So here is what I did:

- I upgraded the RD server console to tgui2.  Its better than it was before but until a refactor of the logging code on the RD Console is done, not a lot of reasons to do more.
- The servers run fine when the temp is below 0C, and stop generating points once they reach 20C.  The server lights show the status as well as the console.
- Once you overheat the servers, they will start sparkling and won't return to normal till the temperature gets below 0C again.
- Yes, you can build more servers for more point generation, however, the scaling factor kills any reason having more than 3 at one time.  Not to mention you need to upgrade the server room the more servers you add as that lonely T1 freezer just barley keeps two servers down in the green.
- I implemented the server heat effect as a component, so now you can make other machines or atoms passively heat up a room!  Will probably come back to it to implement endothermic mode, like for ice blocks or such.

This PR currently enables the servers on default, however, you can disable this entire system by changing  SSresearch.multiserver_calculation  to FALSE.  

## Why It's Good For The Game
These things have had their own little room for more than 6 years and has sit there while people paint the walls with monkey blood.  It makes sense, lore wise, that there would also be a place that all these research points are stored in rather it just generated in the eather. Maybe down the line we could have the points and tech physically stored in the servers so if some idiot blows up the room, someone gets fired for not making disk backups.

Well holding off on that till techweb gets cleaned up.

Currently there shouldn't be any map issues.  I tested it with Box, Meta, Delta, and even Kilo and they all operate the same.   Have had it run 4 hours in the background on the chance I miscalculated the heat values but they still chugging along.

## Changelog
:cl:
 add: Added initial pr for review

/:cl:

### CODE REVIEW

File: code/datums/components/thermo.dm [hunk 1]
```diff
@@ -0,0 +1,78 @@
+ // This is basicly a component that acts like a space heater
+ // Like something thatt emits heat like a flare or absorbes heat like ice
+ // You can set up ranges for when it starts smoking or when it catches fire or
+ //  when it ices over and frezes over, but there is no implmention of sever cold yet
+// Most of it is copyed over from the space_heater code
+/datum/component/thermo
```
Comment (Qustinnus) [inline code/datums/components/thermo.dm:6]: Use autodoc for your component as described here https://github.com/tgstation/tgstation/blob/master/.github/AUTODOC_GUIDE.md

**Comment by MacBlaze1** [comment]
Good PR, the only complaints people should have are "SCIENCE BUFF AAAAAAAAAAAAAAAAAA".

File: code/datums/components/thermo.dm [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+		return PROCESS_KILL
+	var/atom/A = parent
+	var/turf/L = get_turf(A.loc)
+	var/datum/gas_mixture/env
+	if(istype(L))
+		env = L.return_air()
+		// This is from the RD server code.  It works well enough but I need to move over the
+		// sspace heater code so we can caculate power used per tick as well and making this both
+		// exothermic and an endothermic component
+		if(env && env.temperature < targetTemperature)
```
Comment (Qustinnus) [inline code/datums/components/thermo.dm:62]: too much indenting, in most of these cases you can do early returns and not indent so deep

File: code/datums/components/thermo.dm [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+/datum/component/thermo/UnregisterFromParent()
+	STOP_PROCESSING(SSobj, src)
+
+
+
+/datum/component/thermo/proc/get_env_temp()
+	// if we are on and ran though one tick
+	if(on && current_temp >= 0)
+		return current_temp
+	else
```
Comment (Qustinnus) [inline code/datums/components/thermo.dm:30]: you dont need thise else, the if above it returns.

File: code/modules/research/server.dm [hunk 4]
```diff
@@ -4,44 +4,70 @@
...
+	// This allows the servers to rapidly heat up in under 5 min to the shut off point and make it annoying to cool back
+	// down, giving time for RD to fire the guy who shut off the cooler
+
+	var/heating_power = 10000		// Changed the value from 40000.  Just enough for a T1 freezer to keep up with 2 of them
+	var/heating_effecency = 0.25
+	var/temp_tolerance_low = T0C
 	var/temp_tolerance_high = T20C
+	var/temp_tolerance_damage = T0C + 200		// Most CPUS get up to 200C they start breaking.  TODO: Start doing damage to the server?
 	var/temp_penalty_coefficient = 0.5	//1 = -1 points per degree above high tolerance. 0.5 = -0.5 points per degree above high tolerance.
+	var/datum/component/thermo/thermo
```
Comment (Qustinnus) [inline code/modules/research/server.dm:31]: You should not save references to components like this, its generally bad practice and you should be listening to events instead. In this case you would need a OnRefreshParts signal on /obj/machinery/proc/RefreshParts()

File: code/modules/research/server.dm [hunk 5]
```diff
@@ -4,44 +4,70 @@
...
+		server_id = test_id
+
+	name += " [uppertext(num2hex(server_id, -1))]" //gives us a random four-digit hex number as part of the name. Y'know, for fluff.
 	SSresearch.servers |= src
 	stored_research = SSresearch.science_tech
 	var/obj/item/circuitboard/machine/B = new /obj/item/circuitboard/machine/rdserver(null)
 	B.apply_default_parts(src)
-	current_temp = get_env_temp()
+	// The +10 is so the sparks work
+	thermo = LoadComponent(/datum/component/thermo,temp_tolerance_damage + 10,heating_power,heating_effecency)
```
Comment (Qustinnus) [inline code/modules/research/server.dm:52]: not necessary

File: code/modules/research/server.dm [hunk 6]
```diff
@@ -4,44 +4,70 @@
...
+	RefreshParts()
 
 /obj/machinery/rnd/server/Destroy()
+	thermo = null
```
Comment (Qustinnus) [inline code/modules/research/server.dm:56]: not necessary

File: code/modules/research/server.dm [hunk 7]
```diff
@@ -75,123 +121,88 @@
...
-	for(var/obj/machinery/rnd/server/S in no_id_servers)
-		var/num = 1
-		while(!S.server_id)
-			if(num in server_ids)
-				num++
-			else
-				S.server_id = num
-				server_ids += num
-		no_id_servers -= S
+	return thermo.get_env_temp()
```
Comment (Qustinnus) [inline code/modules/research/server.dm:135]: this should be done through the component, not through referencing it.

**Comment by BeeSting12** [comment]
Your changelog is something read by the players so you should probably change it.

File: code/datums/components/thermo.dm [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+		// otherwise we get the temp from the turf
+		var/atom/A = parent
+		var/turf/L = get_turf(A.loc)
+		var/datum/gas_mixture/env
+		if(istype(L))
+			env = L.return_air()
+		return env ? env.temperature : T20C			// env might be null at round start.  This stops runtimes
+
+/datum/component/thermo/proc/enable()
```
Comment (Qustinnus) [inline code/datums/components/thermo.dm:39]: enable and disable might as well be one proc, and disable should be calling STOP_PROCESSING

File: code/datums/components/thermo.dm [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+
+/datum/component/thermo/proc/disable()
+	on = FALSE
+
+/datum/component/thermo/Destroy()
+	return ..()
+
+/datum/component/thermo/process()
+	if(!on)
+		return PROCESS_KILL
```
Comment (Qustinnus) [inline code/datums/components/thermo.dm:53]: why bother doing this if you can just turn it off when you disable it

File: code/modules/research/server.dm (hunk 7 above)
Comment (WarlockD) [reply code/modules/research/server.dm:135]: Its it cost effective to keep calling GetComponent for it though?  Whole reason I am deref is the component constantly updates the current temp though its process so I would of thought it be more efficient to just pull that data from that.

File: code/datums/components/thermo.dm (hunk 9 above)
Comment (WarlockD) [reply code/datums/components/thermo.dm:53]: Threading habits mostly.  Keep seeing process as a seperate thread and want to kill it in the thread:P

File: code/modules/research/server.dm (hunk 7 above)
Comment (Qustinnus) [reply code/modules/research/server.dm:135]: the thing is that you shouldn't be touching the component at all; the idea of components on TG is that they are generic behavior you can add to an object without having to interact with the component outside of making callbacks.

Everything should be handled through signals (And those signals should not be used like proc calls; they should essentialy be events the component is listening for like for example when the object its attached to is hit)

**Comment by ExcessiveUseOfCobblestone** [comment]
Goes against [R&D Design Doc](https://hackmd.io/A6KYwyAASZinVQJdqX1AUg)

**Comment by BeeSting12** [comment]
It doesn't have to go against, this just makes it possible to sabotage research. Remove the more points scaling and just make it a flat rate that drops if the servers aren't kept in the cold. Making research impossible to sabotage is dumb imo.

**Comment by ExcessiveUseOfCobblestone** [comment]
`Flat rate point generation, from start of round and ticking upwards.`

Being able to scale pointgain with amount of servers is not flat.

The suggestion from scaling with temp as well as scaling with amount of servers to just with amount of servers is still not flat. If you have an issue with it, ask oranges to change the design doc.

**Comment by optimumtact** [comment]
I was quite clear

**Comment by WarlockD** [comment]
You can't scale much, two servers produce the exact as the static number, and decrease if someone kills the heater.  It would be simple to fix it so more servers don't give you more points or the heat generation becomes so nightmarish, its easier to build a second SM.

But since it's just closed, is it better for me to just make a pr to delete this all together?  It would free up map space that is dead in the center of RD.

**Pull request closed without merging**